caravel/controllers/custom_fields.py

```diff
@@ -19,6 +21,26 @@ def __call__(self, form, field):
         if not field.data or not field.data.valid:
             raise ValidationError("Invalid credential.")
 
+class MatchesPrincipal(threading.local):
+    """Ensures that only the existing Principal is used."""
+
+    def __init__(self):
+        self.principal = None
+
+    def __call__(self, form, field):
+        """Ensure that only the given user is used for authentication."""
+
+        if not self.principal:
+            return
+
+        if not field.data or not field.data.can_act_as(self.principal):
+            if not field.data:
+                raise ValidationError("Please enter an email.")
+            elif field.data.email == self.principal.email:
```

Comment:
If someone is trying to log in with their CNetID, the field.data will be a Principal representing that email. In that case, we will already have passed the can_act_as(...) check.